File: android/src/main/java/com/documentscanner/Options.kt
package com.documentscanner

import com.facebook.react.bridge.ReadableMap

/**
 * Base options shared by both Scan and Process operations.
 * Allows parsing common properties from a React Native ReadableMap.
 */
abstract class BaseOptions(
    val quality: Double,
    val format: String,
    val filter: String,
    val includeBase64: Boolean,
    val includeText: Boolean,
    val textVersion: Int
) {
    /**
     * Secondary constructor to parse common options from a dictionary.
     * Delegates to the primary constructor with validated values.
     */
    constructor(options: ReadableMap?, defaultIncludeText: Boolean) : this(
        quality = parseQuality(options),
        format = parseFormat(options),
        filter = parseFilter(options),
        includeBase64 = parseBoolean(options, "includeBase64", false),
        includeText = parseBoolean(options, "includeText", defaultIncludeText),
        textVersion = parseTextVersion(options)
    )

    companion object {
        /** Validates and returns the compression quality [0.1, 1.0]. */
        private fun parseQuality(options: ReadableMap?): Double {
            val value = if (options?.hasKey("quality") == true) options.getDouble("quality") else 1.0
            return value.coerceIn(0.1, 1.0)
        }

        /** Whitelists and returns the output image format. */
        private fun parseFormat(options: ReadableMap?): String {
            val value = options?.getString("format") ?: "jpg"
            return if (value == "png") "png" else "jpg"
        }

        /** Whitelists and returns the post-processing filter. */
        private fun parseFilter(options: ReadableMap?): String {
            val value = options?.getString("filter") ?: "color"
            val validFilters = listOf("color", "grayscale", "monochrome", "denoise", "sharpen", "ocrOptimized")
            return if (validFilters.contains(value)) value else "color"
        }

        /** Safely parses a boolean value with a default fallback. */
        private fun parseBoolean(options: ReadableMap?, key: String, default: Boolean): Boolean {
            return if (options?.hasKey(key) == true) options.getBoolean(key) else default
        }

        /** Validates and returns the OCR engine version [1, 2]. */
        private fun parseTextVersion(options: ReadableMap?): Int {
            val value = if (options?.hasKey("textVersion") == true) options.getInt("textVersion") else 2
            return if (value == 1) 1 else 2
        }
    }
}

/**
 * Strongly-typed options for the scanDocuments() method.
 */
class ScanOptions(options: ReadableMap?, fallbackPageCount: Int) : BaseOptions(options, false) {
    /** Maximum pages to scan, clamped to [0, 100]. */
    val maxPageCount: Int = parseMaxPages(options, fallbackPageCount)

    companion object {
        fun from(options: ReadableMap?): ScanOptions = ScanOptions(options, 0)

        private fun parseMaxPages(options: ReadableMap?, fallback: Int): Int {
            val value = if (options?.hasKey("maxPageCount") == true) options.getInt("maxPageCount") else fallback
            return value.coerceIn(0, 100)
        }
    }
}

/**
 * Strongly-typed options for the processDocuments() method.
 */
class ProcessOptions(options: ReadableMap?) : BaseOptions(options, true) {
    /** List of image URIs or Base64 strings to process. */
    val images: List<String> = parseImages(options)

    companion object {
        fun from(options: ReadableMap?): ProcessOptions = ProcessOptions(options)

        private fun parseImages(options: ReadableMap?): List<String> {
            val images = mutableListOf<String>()
            if (options?.hasKey("images") == true) {
                val array = options.getArray("images")
                if (array != null) {
                    for (i in 0 until array.size()) {
                        array.getString(i)?.let { images.add(it) }
                    }
                }
            }
            return images
        }
    }
}

File: ios/Options.swift
import Foundation
import UIKit

/**
 Base options shared by both Scan and Process operations.
 Contains configuration for image output and OCR processing.
 */
public class BaseOptions {
    public let quality: CGFloat
    public let format: String
    public let filter: String
    public let includeBase64: Bool
    public let includeText: Bool
    public let textVersion: Int

    init(quality: CGFloat, format: String, filter: String, includeBase64: Bool, includeText: Bool, textVersion: Int) {
        self.quality = quality
        self.format = format
        self.filter = filter
        self.includeBase64 = includeBase64
        self.includeText = includeText
        self.textVersion = textVersion
    }

    /**
     Reads an integer option from the bridge dictionary.
     Covers NSNumber, Int, and Double — all three bridging representations that can appear
     depending on whether the call comes through the old bridge or JSI (new arch).
     */
    static func intOption(from dictionary: [String: Any]?, key: String, fallback: Int) -> Int {
        guard let raw = dictionary?[key] else { return fallback }
        if let n = raw as? NSNumber { return n.intValue }
        if let i = raw as? Int { return i }
        if let d = raw as? Double { return Int(d) }
        return fallback
    }

    /**
     Convenience initializer to parse common options from a dictionary.
     - Parameters:
       - dictionary: Raw options dictionary.
       - defaultIncludeText: Default value for includeText (Scan defaults to false, Process to true).
     */
    init(from dictionary: [String: Any]?, defaultIncludeText: Bool) {
        /* Quality: Clamp [0.1, 1.0] */
        let q = dictionary?["quality"] as? CGFloat ?? 1.0
        self.quality = max(0.1, min(1.0, q))

        /* Format: whitelist [jpg, png] */
        let f = dictionary?["format"] as? String ?? "jpg"
        self.format = (f == "png") ? "png" : "jpg"

        /* Filter: whitelist supported types */
        let filterInput = dictionary?["filter"] as? String ?? "color"
        let validFilters = ["color", "grayscale", "monochrome", "denoise", "sharpen", "ocrOptimized"]
        self.filter = validFilters.contains(filterInput) ? filterInput : "color"

        self.includeBase64 = dictionary?["includeBase64"] as? Bool ?? false
        self.includeText = dictionary?["includeText"] as? Bool ?? defaultIncludeText

        /* Text Version: allow [1, 2]. */
        let rawVersion = BaseOptions.intOption(from: dictionary, key: "textVersion", fallback: 2)
        self.textVersion = (rawVersion == 1) ? 1 : 2
    }
}

/**
 Strongly-typed representation of scan options.
 Parses the raw dictionary from React Native and provides defaults.
 */
public class ScanOptions: BaseOptions {
    public let maxPageCount: Int

    /**
     Initializes ScanOptions from a raw dictionary.
     - Parameters:
       - dictionary: The options dictionary from React Native.
       - fallbackPageCount: Default page count if not specified.
     */
    public init(from dictionary: [String: Any]?, fallbackPageCount: Int) {
        /* Max Page Count: Clamp [0, 100]. 0 = unlimited. */
        let rawMax = BaseOptions.intOption(from: dictionary, key: "maxPageCount", fallback: fallbackPageCount)
        self.maxPageCount = max(0, min(100, rawMax))

        super.init(from: dictionary, defaultIncludeText: false)
    }
}

/**
 Strongly-typed representation of process options.
 Parses the raw dictionary from React Native for processDocuments.
 */
public class ProcessOptions: BaseOptions {
    public let images: [String]

    /**
     Initializes ProcessOptions from a raw dictionary.
     - Parameter dictionary: The options dictionary from React Native.
     - Returns: nil if 'images' array is missing.
     */
    public init?(from dictionary: [String: Any]?) {
        guard let dict = dictionary,
              let images = dict["images"] as? [String] else {
            return nil
        }
        self.images = images
        super.init(from: dictionary, defaultIncludeText: true)
    }
}
